internal/app/client/client.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"flag"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"connectrpc.com/conformance/internal/app"
+	"connectrpc.com/conformance/internal/gen/proto/connect/connectrpc/conformance/v1alpha1/conformancev1alpha1connect"
+	v1alpha1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1alpha1"
+	"connectrpc.com/connect"
+	"golang.org/x/net/http2"
+)
+
+// Wrapper defines a wrapper around a client which can invoke requests based on a ClientCompatRequest.
+type Wrapper interface {
+	// Invoke invokes a request according to a given config
+	Invoke(context.Context, *v1alpha1.ClientCompatRequest) (*v1alpha1.ClientCompatResponse, error)
+}
+
+// Run runs the server according to server config read from the 'in' reader.
+func Run(ctx context.Context, _ []string, inReader io.ReadCloser, outWriter io.WriteCloser) error {
+	json := flag.Bool("json", false, "whether to use the JSON format for marshaling / unmarshaling messages")
+
+	flag.Parse()
+
+	// Read the server config from  the in reader
+	data, err := io.ReadAll(inReader)
+	if err != nil {
+		return err
+	}
+
+	codec := app.NewCodec(*json)
+
+	req := &v1alpha1.ClientCompatRequest{}
+	if err := codec.Unmarshal(data, req); err != nil {
+		return err
+	}
+
+	resp, err := invoke(ctx, req)
+	if err != nil {
+		return err
+	}
+	bytes, err := codec.Marshal(resp)
+	if err != nil {
+		return err
+	}
+	if _, err := outWriter.Write(bytes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func invoke(ctx context.Context, req *v1alpha1.ClientCompatRequest) (*v1alpha1.ClientCompatResponse, error) {
+	var scheme string
+	if req.ServerTlsCert != nil {
+		scheme = "https://"
+	} else {
+		scheme = "http://"
+	}
+	urlString := scheme + net.JoinHostPort(req.Host, strconv.FormatUint(uint64(req.Port), 10))
+	serverURL, err := url.ParseRequestURI(urlString)
+	if err != nil {
+		return nil, errors.New("invalid url: %s" + urlString)
+	}
+
+	var transport http.RoundTripper
+	switch req.HttpVersion {
+	case v1alpha1.HTTPVersion_HTTP_VERSION_1:
+		transport = &http.Transport{}
+	case v1alpha1.HTTPVersion_HTTP_VERSION_2:
+		transport = &http2.Transport{
```

Comment:
This seems like it will be a problem if/when we have 100s of test cases as this creates a new `http.Transport` for every request.

Maybe add a TODO to cache/re-use transports instead of creating a new one for every request?